### DISCUSSION TITLE
very slow work with 400 rows table

### DISCUSSION THREAD

**Discussion by vityah1** [django-commons/django-unicorn#320] [Q&A]
Hello

This is normal when django-unicorn with table on 400 records works very slow?

I wait 10 sec until see any reaction on click

Thanks

**Comment by adamghill** [comment]
`Unicorn` is easy to start with, but the trade-off is that sometimes it needs to be tweaked for certain use-cases. For a lot of data, you might need to tweak some things for acceptable performance. Suggestions of where I would look:

- If the click is only updating a portion of the page, partial update might be helpful: https://www.django-unicorn.com/docs/partial-updates/
- Exclude any data from being rendered for JavaScript that isn't absolutely necessary to prevent unnecessary serialization: https://www.django-unicorn.com/docs/advanced/#javascript-exclude

If you can fork the repo and add some code to the example project or have some example code you can put into a public repo, I might be able to give you more specific help to optimize performance (or make updates to `Unicorn` which would benefit everybody).

**Reply by vityah1** [reply]
Thanks for answer.

I show some of my code here. if you check my code i will be very grateful for your help

This is template with table with 400-500 rows.
[phone.html.zip](https://github.com/adamghill/django-unicorn/files/7762298/phone.html.zip)

This is my code
[phone.py.zip](https://github.com/adamghill/django-unicorn/files/7762299/phone.py.zip)

I thinks this is typical CRUDe like code for some dictionaries.

May be you can say me what I need change for work faster?

Thanks.

**Comment by vityah1** [comment]
So, not any way to speed up work ?
I dont know how I can use partial update if I need reaction on click every of 400 rows.
Do you can advise something  me about it?

**Reply by adamghill** [reply]
Thanks for the code you sent me. The reason I ask for a fork or a repo is so I don't need to integrate your code into my tests manually. For example, I'm also unsure what your `Phone` model fields are like. I just used this for my testing:

```
class Phone(models.Model):
    deleted = models.IntegerField(default=0)
    iptel = models.CharField(blank=True, null=True, max_length=1024)
    cdep = models.CharField(blank=True, null=True, max_length=1024)
    c_sti = models.CharField(blank=True, null=True, max_length=1024)
    posada = models.CharField(blank=True, null=True, max_length=1024)
    cpib = models.CharField(blank=True, null=True, max_length=1024)
    kab = models.CharField(blank=True, null=True, max_length=1024)
    tel = models.CharField(blank=True, null=True, max_length=1024)
```

Here's a GIF of the response time on my local when I click a link in the list of 400.

![Dec-28-2021 07-55-11](https://user-images.githubusercontent.com/317045/147568581-98e5006a-c2d4-40c1-bd94-4ce1f37c4303.gif)

However, notice the response size of 150 KB-200 KB? If your model has a lot of data in it that size will become very large. That might be one reason the response time is slow. https://www.django-unicorn.com/docs/advanced/#javascript-exclude can be used to prevent parts of the model from being serialized and sent back and forth on every request/response.

Do you run into the same issues when it's only 5 instead of 400 results? Another approach could be to paginate the results.

There might be other solutions as well, but it's challenging to debug without having all the code to replicate the problem. Let me know if you continue running into problems and I'll try to help as best as I can!

**Comment by vityah1** [comment]
Very thanks for Your answer

I attach video with my work responces
May be this help understant situation

https://user-images.githubusercontent.com/54395237/147571335-6de38ee7-0a39-471c-b8eb-ab93fe3b53d1.mp4

I see that very slow work in the backend

**Reply by adamghill** [reply]
Interesting! I see the `print(f"py: get phone: {id}")` in your component. Can you add some other print statements after `phone = Phone.objects.get(id=id)`? I wouldn't expect an `id` lookup to be slow because it gets indexed by default in Django, but maybe something else is going on?

Other options that might give you some insights:
- use the `timeit` module and print out the elapsed time between parts of your code (see https://www.analyticsvidhya.com/blog/2020/09/timeit-in-python/ for an example)
- if it is a database issue, `django-debug-toolbar` and the SQL panel might be useful to see why it's slow

Let me know what you find!

**Reply by vityah1** [reply]
I made screenshot with timing process
I see that all operation in func get_phone do quick. But after that  - responce from server go very slow.
After last print with open modal and POST I wait 5-6 s

![timing](https://user-images.githubusercontent.com/54395237/147665104-c2812973-70f3-4f55-85f4-82c319862656.png)

**Reply by adamghill** [reply]
Can you send me the data for that phone record? I'm going to see if serialization is super slow for non-Latin characters or something, but that will be easier if you can give me the data. :)

**Reply by vityah1** [reply]
Thanks for answer
Yes, of couse
I add file with insert statements

[teldpaip.sql.zip](https://github.com/adamghill/django-unicorn/files/123456/teldpaip.sql.zip) /deleted/

the codepage is win1251 or cp1251

**Reply by vityah1** [reply]
Hi
Have a nice day
Happy 2022 NY :)
Do you find some solution for this situation?
Thanks

**Reply by adamghill** [reply]
Sorry, I finally inserted your SQL and tested on my local. Everything looks pretty quick to me? You can see the Python print statements and then the console log is almost instantaneous.

![Kapture 2022-01-15 at 08 40 40](https://user-images.githubusercontent.com/317045/149624083-fb04afeb-d025-4af4-bcef-66bef42937ac.gif)

I'm only testing on local and with a SQLite database, though.

**Reply by vityah1** [reply]
Thanks for answer.
So, I don't have any idea why it work so slowly for me. 
As I see. Database work quick. But python + django+unicorn on windows platform work slow.
